Fix all traffic widget label when there is only a single row of data (#2660)

johnPhillips opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #2660

## Relevant technical choices

<!-- Please describe your changes. -->

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

eugene-manuilov reviewed (approved):
:+1:

aaemnnosttv commented on assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/UserDimensionsPieChart.js:
```diff
@@ -292,11 +292,17 @@ export default function UserDimensionsPieChart( { dimensionName, dimensionValue,
 		options.pieSliceTextStyle.color = 'transparent';
 	}
 
+	if ( report?.[0]?.data?.rows?.length === 1 ) {
+		// When there is only one row, the chart will add a label which we need to hide - see issue #2660
+		options.pieSliceText = 'none';
+	}
```
@johnPhillips does this condition need to be separate than the one above? They both seem to cater for the single slice case and do different things. Is there any reason to change the `pieSliceTextStyle.color` in this case if we're changing the `pieSliceText` to `none` ?

johnPhillips replied:
I spoke to @eugene-manuilov who wrote the other condition here and it seems they are both attacking the same problem. I've combined them.

aaemnnosttv reviewed (commented):
Code looks good! Sending back to add a story for this state per our convo.

aaemnnosttv reviewed (commented):
Thanks @johnPhillips – the added story is really helpful.

The only thing I see now is the alignment of the text looks different than the image in the QA Brief. The component in Storybook shows the text somewhat below center. Did this change with the new V2 chart component maybe? (cc: @tofumatt )

**QA Brief**
![original](https://user-images.githubusercontent.com/2470911/107524542-cb582d80-6bad-11eb-9ba5-50cdede16067.png)

**Storybook**
![image](https://user-images.githubusercontent.com/1621608/109636605-737d6880-7b54-11eb-997a-304d1f28b2ff.png)

johnPhillips commented:
@aaemnnosttv  Looks like as part of the GoogleChartV2 work we have added the `googlesitekit-cta-link googlesitekit-widget--analyticsAllTraffic__legend-slice` at the bottom. It has a negative top margin, which changes the height of the parent, meaning the absolutely positioned title appears off-center. I added a fix 👍

aaemnnosttv commented on assets/sass/widgets/_googlesitekit-widget-analyticsAllTraffic.scss:
```diff
@@ -167,7 +167,10 @@
 		text-align: center;
 		text-transform: uppercase;
-		top: 50%;
+		// The -6px margin-top on the legend affects the parent height
+		// so we have to factor it in here to center the title
+		top: calc(50% - 6px);
```
I'm not sure this is the right solution here. Looking at the component, the title seems to be intended to be a child of the chart component and the legend below:
https://github.com/google/site-kit-wp/blob/0075f6de4035b67d5830ada399bbc8cdca9c5a37/assets/js/modules/analytics/components/dashboard/DashboardAllTrafficWidget/UserDimensionsPieChart.js#L387-L410
However, due to an implementation detail of the chart being output as a fragment with the chart followed by the children, this results in some unexpected output where the title and legend are siblings of each other which causes the legend to influence the height of the pie chart when it really shouldn't.

I would suggest we update `GoogleChartV2` to provide a wrapping `div` instead so that the title is always relative to the area of the chart.

<details>
<summary>This worked well in my testing:</summary>

![Kapture 2021-03-03 at 13 00 25](https://user-images.githubusercontent.com/1621608/109796928-6ed2b600-7c21-11eb-8324-ab73f90ed42a.gif)

</details>
@tofumatt any reason we can't do something like this?

<details>
<summary>git diff</summary>

```diff
diff --git a/assets/js/components/GoogleChartV2.js b/assets/js/components/GoogleChartV2.js
index cfef6c26d..4e2fcc4fc 100644
--- a/assets/js/components/GoogleChartV2.js
+++ b/assets/js/components/GoogleChartV2.js
@@ -136,13 +136,19 @@ export default function GoogleChartV2( props ) {
 	}
 
 	return (
-		<Fragment>
+		<div
+			className={ classnames(
+				'googlesitekit-chart-v2',
+				`googlesitekit-chart-v2--${ chartType }`,
+				className
+			) }
+		>
 			<Chart
 				chartEvents={ combinedChartEvents }
 				chartType={ chartType }
 				className={ classnames(
-					'googlesitekit-chart-v2',
-					`googlesitekit-chart-v2--${ chartType }`,
+					'googlesitekit-chart-v2__chart',
+					`googlesitekit-chart-v2__chart--${ chartType }`,
 					className
 				) }
 				loader={ loader }
@@ -170,7 +176,7 @@ export default function GoogleChartV2( props ) {
 				{ ...otherProps }
 			/>
 			{ children }
-		</Fragment>
+		</div>
 	);
 }
 
diff --git a/assets/sass/components/global/_googlesitekit-chart.scss b/assets/sass/components/global/_googlesitekit-chart.scss
index 03294e911..7d083cf0b 100644
--- a/assets/sass/components/global/_googlesitekit-chart.scss
+++ b/assets/sass/components/global/_googlesitekit-chart.scss
@@ -24,6 +24,10 @@
 	}
 }
 
+.googlesitekit-chart-v2 {
+	position: relative;
+}
+
 .googlesitekit-chart-v2--PieChart svg > g > g {
 	pointer-events: none;
 }
```
</details>

tofumatt replied:
No reason at all that shouldn't work, though we'll want to adjust a single CSS rule we have that has Hyper-Selector-Specificity™: https://github.com/google/site-kit-wp/blob/0075f6de4035b67d5830ada399bbc8cdca9c5a37/assets/sass/widgets/_googlesitekit-widget-analyticsAllTraffic.scss#L144-L150

I think it's a matter of adding another `> div` to make up for what `Chart` renders. We have to be really specific there to override and target specific SVG styles, if I recall correctly. (Alternatively, add a `googlesitekit-chart-v2__inner` to `Chart`'s `className` or something and use that.)

(Also: I'd strip out the `className` from the `Chart` component entirely there rather than duplicate it in the `div` and `Chart` component, but I think that might've been done by mistake 😅)

TL;DR is yes replacing the `Fragment` with a `<div>` should be fine 👍🏻

johnPhillips replied:
Thanks @tofumatt 
Right, let me give this a whirl...

aaemnnosttv reviewed (changes requested):
Thanks @johnPhillips – the CSS change works but I don't think it's actually centered due to the height of the legend also affecting the total height. The good news is I don't think we need to change the CSS to fix it. See my comments below.

cc: @tofumatt

aaemnnosttv commented on .storybook/storybook-data.js:
```diff
@@ -65,7 +65,7 @@ module.exports = [
 				hierarchyRootSeparator: '|',
 				hierarchySeparator: {},
-				readySelector: '.googlesitekit-chart-v2 > div[style="position: relative;"]',
+				readySelector: '.googlesitekit-chart-v2',
```
I think the intention here is to wait for the _inner_ selector – the one generated by a rendered chart, whereas `googlesitekit-chart-v2` will be present on all renderings of the chart, (loading, etc) so this may cause problems with the reliability of the test.

johnPhillips replied:
Ah yes you are right, my bad. I've addressed this (hopefully!)

aaemnnosttv reviewed (commented):
Thanks @johnPhillips – the title looks great now and is no longer affected by the legend 👍 

I left one comment about the `readySelector` change but also VRT is failing (maybe related?)

aaemnnosttv reviewed (approved):
Super, thanks @johnPhillips !
